akka-cluster-tools/src/main/scala/akka/cluster/singleton/ClusterSingletonManager.scala

```diff
@@ -422,6 +452,16 @@ class ClusterSingletonManager(
     removed = removed filter { case (_, deadline) ⇒ deadline.hasTimeLeft }
   }
 
+  // for CoordinatedShutdown
+  val coordShutdown = CoordinatedShutdown(context.system)
+  val memberExitingProgress = Promise[Done]()
+  coordShutdown.addTask(CoordinatedShutdown.PhaseClusterExiting)(() ⇒
+    memberExitingProgress.future)
+  coordShutdown.addTask(CoordinatedShutdown.PhaseClusterExiting) { () ⇒
+    implicit val timeout = Timeout(coordShutdown.timeout(CoordinatedShutdown.PhaseClusterExiting))
+    self.ask(SelfExiting).mapTo[Done]
+  }
```

Comment:
Register just one task here and `scala.concurrent.Future.firstCompletedOf()` the futures instead?